### PR TITLE
prove `invokeSchedContext_corres`

### PR DESCRIPTION
**Pull request by mktnk3**
- also includes proofs for `schedContextBindNtfn` and `schedContextYieldTo_corres`
- resolves the issue of IPC buffer preconditions for `setConsumed_corres`

### CODE REVIEW

File: proof/refine/ARM/SchedContextInv_R.thy [hunk 1]
```diff
@@ -371,18 +374,393 @@ lemma decode_sc_ctrl_inv_corres:
...
+     apply wpsimp+
+   apply (fastforce simp: obj_at_simps sc_ntfn_sc_at_def is_ntfn is_sc_obj valid_obj_def)
+  apply clarsimp
+  done
+
+crunches tcb_sched_action, complete_yield_to, reschedule_required, sched_context_resume
+  for in_user_frame[wp]: "in_user_frame buf"
+  (simp: crunch_simps wp: crunch_wps ignore: set_tcb_obj_ref)
+
+lemma
```
Comment (michaelmcinerney) [inline proof/refine/ARM/SchedContextInv_R.thy:411]: These can't be crunched?

Comment (michaelmcinerney) [reply proof/refine/ARM/SchedContextInv_R.thy:411]: I guess not, if you do some crunching immediately after. I wonder if it's worth investigating

File: proof/refine/ARM/SchedContextInv_R.thy [hunk 2]
```diff
@@ -371,18 +374,393 @@ lemma decode_sc_ctrl_inv_corres:
...
+  apply (fastforce simp: obj_at'_def projectKOs inQ_def)
+  done
+
+lemma sc_yield_from_update_sc_tcb_sc_at[wp]:
+  "set_sc_obj_ref sc_yield_from_update scp ptr \<lbrace>sc_tcb_sc_at P scp\<rbrace>"
+  apply (wpsimp wp: update_sched_context_wp)
+  by (clarsimp simp: sc_tcb_sc_at_def obj_at_def)
+
+lemma schedContextYieldTo_corres:
+  "corres dc ((case_option \<top> in_user_frame buf and einvs) and (\<lambda>s. bound_yt_tcb_at ((=) None) (cur_thread s) s \<and>
```
Comment (michaelmcinerney) [inline proof/refine/ARM/SchedContextInv_R.thy:451]: Maybe we should decide on a way to indent the guards and functions in `corres` rules, or at least narrow it down to a couple options

File: proof/refine/ARM/SchedContextInv_R.thy [hunk 3]
```diff
@@ -371,18 +374,393 @@ lemma decode_sc_ctrl_inv_corres:
...
+                    clarsimp simp: obj_at_def sc_tcb_sc_at_def is_sc_obj
+                            elim!: valid_sched_context_size_objsI dest!: invs_valid_objs)
+              apply (fastforce simp: valid_obj_def obj_at_def valid_sched_context_def)
+             apply (intro conjI impI; clarsimp simp: is_tcb pred_tcb_at_def obj_at_def)
+              apply (erule (1) valid_sched_context_size_objsI)+
+            apply (clarsimp simp: invs'_def valid_pspace'_def valid_objs'_valid_tcbs')
+            apply (clarsimp simp: obj_at'_def projectKOs opt_map_red)
+           apply (rule_tac Q="\<lambda>rv. (case buf of None \<Rightarrow> \<lambda>_. True | Some x \<Rightarrow> in_user_frame x) and einvs and
+                                     sc_yf_sc_at ((=) None) scp and
+                                     (\<lambda>s. bound_yt_tcb_at ((=) None) (cur_thread s) s \<and>
```
Comment (michaelmcinerney) [inline proof/refine/ARM/SchedContextInv_R.thy:658]: The formatting in these `hoare_strengthen_post`s could be cleaned up a bit

**Review by michaelmcinerney — APPROVED** [review]
Some nice cleanup along the way too

File: proof/refine/ARM/SchedContextInv_R.thy [hunk 4]
```diff
@@ -371,18 +374,393 @@ lemma decode_sc_ctrl_inv_corres:
...
+     (ntfn_at' ntfnp and sc_at' scp)
+     (sched_context_bind_ntfn scp ntfnp) (schedContextBindNtfn scp ntfnp)"
+  unfolding sched_context_bind_ntfn_def schedContextBindNtfn_def
+  apply (clarsimp simp: update_sk_obj_ref_def bind_assoc)
+  apply (rule corres_guard_imp)
+    apply (rule corres_split[OF getNotification_corres])
+      apply (rule corres_guard_imp)
+        apply (rule corres_split[OF setNotification_corres])
+           apply (clarsimp simp: ntfn_relation_def split: Structures_A.ntfn.splits)
+          apply (fold updateSchedContext_def)
```
Comment (corlewis) [inline proof/refine/ARM/SchedContextInv_R.thy:390]: I guess it's not worth changing `schedContextBindNtfn` to actually use `updateSchedContext`?

File: proof/refine/ARM/SchedContextInv_R.thy (hunk 2 above)
Comment (corlewis) [reply proof/refine/ARM/SchedContextInv_R.thy:451]: Yeah, agreed. It would be nice to not have to think about every time it comes up.

Comment (corlewis) [reply proof/refine/ARM/SchedContextInv_R.thy:451]: In this specific case, the indentation within the preconditions seems a bit off as well and I think that some of the bracketing can be simplified.

File: proof/refine/ARM/SchedContextInv_R.thy [hunk 5]
```diff
@@ -371,18 +374,393 @@ lemma decode_sc_ctrl_inv_corres:
...
+                             cong: conj_cong imp_cong if_cong)
+             apply (rule context_conjI;
+                    clarsimp simp: obj_at_def sc_tcb_sc_at_def is_sc_obj
+                            elim!: valid_sched_context_size_objsI dest!: invs_valid_objs)
+              apply (fastforce simp: valid_obj_def obj_at_def valid_sched_context_def)
+             apply (intro conjI impI; clarsimp simp: is_tcb pred_tcb_at_def obj_at_def)
+              apply (erule (1) valid_sched_context_size_objsI)+
+            apply (clarsimp simp: invs'_def valid_pspace'_def valid_objs'_valid_tcbs')
+            apply (clarsimp simp: obj_at'_def projectKOs opt_map_red)
+           apply (rule_tac Q="\<lambda>rv. (case buf of None \<Rightarrow> \<lambda>_. True | Some x \<Rightarrow> in_user_frame x) and einvs and
```
Comment (corlewis) [inline proof/refine/ARM/SchedContextInv_R.thy:656]: It doesn't make much of a difference, but I guess that these `hoare_strengthen_posts` could use `case_option` in the same way the lemma preconditions do.

File: proof/refine/ARM/SchedContextInv_R.thy [hunk 6]
```diff
@@ -371,18 +374,393 @@ lemma decode_sc_ctrl_inv_corres:
...
+           and current_time_bounded 2)
+          (invs' and sch_act_simple and valid_sc_inv' sc_inv')
           (invoke_sched_context sc_inv)
           (invokeSchedContext sc_inv')"
   apply (simp add: invoke_sched_context_def invokeSchedContext_def)
-  (* most of the next layer down should go into SchedContext_R, because some of these are
-     reused in Finalise and IpcCancel *)
-  sorry
+  apply (cases sc_inv; cases sc_inv'; clarsimp simp: cap_relation_def)
+      defer
```
Comment (corlewis) [inline proof/refine/ARM/SchedContextInv_R.thy:717]: Do you still need this `defer`? Looking at the proof of the final subgoal, it doesn't seem to be doing anything special now that would require it to be left until the end.

**Review by corlewis — APPROVED** [review]
Great work, these are some big functions to have gotten through!

File: proof/refine/ARM/SchedContextInv_R.thy (hunk 2 above)
Comment (mktnk3) [reply proof/refine/ARM/SchedContextInv_R.thy:451]: The brackets are necessary for the correct pattern matching.

File: proof/refine/ARM/SchedContextInv_R.thy (hunk 5 above)
Comment (mktnk3) [reply proof/refine/ARM/SchedContextInv_R.thy:656]: These big strengthening patterns are typically copied over from the output pane, and this is how Isabelle prints it.

File: proof/refine/ARM/SchedContextInv_R.thy (hunk 6 above)
Comment (mktnk3) [reply proof/refine/ARM/SchedContextInv_R.thy:717]: I should remove this. This was for moving `setConsumed` case to the last to deal with it later.

File: proof/invariant-abstract/DetSchedSchedule_AI.thy [hunk 7]
```diff
@@ -18406,6 +18406,15 @@ lemma sched_context_bind_ntfn_valid_sched_pred_strong[wp]:
   unfolding sched_context_bind_ntfn_def
   by wpsimp
 
+crunches sched_context_cancel_yield_to
+  for sc_tcb_sc_at[wp]: "sc_tcb_sc_at P scp"
```
[the remark this answers is not in the document]
Comment (mktnk3) [reply proof/invariant-abstract/DetSchedSchedule_AI.thy]: I tried that initially but it needs to call an exisiting lemma that is not generalised. And changing that one would incur some proof change that I thought wasn’t quite worth going through.

File: proof/refine/ARM/SchedContextInv_R.thy (hunk 1 above)
Comment (mktnk3) [reply proof/refine/ARM/SchedContextInv_R.thy:411]: I did try initially and didn’t work.

File: proof/refine/ARM/KHeap_R.thy [hunk 8]
```diff
@@ -3607,6 +3607,43 @@ lemma bound_sc_tcb_at_cross:
...
+  apply (erule (1) pspace_dom_relatedE)
+  apply (erule (1) obj_relation_cutsE, simp_all)
+   apply (clarsimp simp: st_tcb_at'_def obj_at'_def other_obj_relation_def tcb_relation_def
+                  split: Structures_A.kernel_object.split_asm if_split_asm)+
+  apply fastforce
+  done
+
+(* RT FIXME : Move to InvariantsPre_AI *)
+lemma neq_None_bound: "((\<noteq>) None) = bound"
+  using cases_simp_option option.simps(3) by metis
```
[the remark this answers is not in the document]
Comment (mktnk3) [reply proof/refine/ARM/KHeap_R.thy]: Moved. Not sure about `[simp]` and I didn’t add that. It will require some proof updates, I think, so maybe later.

**Pull request merged**